Mathlib/SetTheory/Ordinal/Clubs.lean

```diff
@@ -0,0 +1,243 @@
+    (∀ i : Ordinal.{u}, (hi : i < ω) → P (f i hi) (f (i + 1) (omega_isLimit.2 i hi)))
+    ∧ (∀ i j, (hi : i < ω) → (hj : j < ω) → (i < j) → f i hi < f j hj)
+    ∧ r < f 0 omega_pos := by
+  have oLim : o.IsLimit := ⟨opos.ne.symm, fun a alto ↦ (hP a alto).casesOn fun r hr ↦
+    lt_of_le_of_lt (succ_le_iff.mpr hr.2.1) hr.1⟩
+  let H₂ : (p : Ordinal) → p < ω → (Iio o) → (Iio o) := fun p _ fp ↦ by
+    let C := choose (hP fp fp.2)
+    have hC := (choose_spec (hP fp fp.2)).1
+    exact ⟨C, hC⟩
+  let H₃ : (w : Ordinal) → w < ω → w.IsLimit → ((o' : Ordinal) → o' < w → (Iio o)) → (Iio o) :=
+    fun w _ _ _ ↦ ⟨0, oLim.pos⟩
+  let f : Π p < ω, Iio o := @boundedLimitRec' (α := Iio o) ω omega_isLimit
+    ⟨r + 1, oLim.succ_lt rlto⟩ H₂ H₃
+  use f
+  constructor <;> try constructor
+  · intro n hn
+    simp [f]
+    generalize_proofs _ pf
+    exact (choose_spec pf).2.2
+  · have aux : ∀ i : Ordinal.{u}, (hi : i < ω) → f i hi < f (i + 1) (omega_isLimit.2 i hi) := by
+      intro i hi
+      simp [f, H₂]
+      generalize_proofs _ _ _ pf
+      exact (choose_spec pf).casesOn fun _ x ↦ x.casesOn fun x _ ↦ x
+    exact monotone_of_succ_lt_omega f aux
+  simp [f]
+
+theorem exists_omega_seq_succ_prop_pos (onelto : 1 < o) {P : Ordinal → Ordinal → Prop}
+    (hP : ∀ p < o, 0 < p → ∃ q < o, (p < q ∧ P p q)) {r} (rlto : r < o) : ∃ f : Π p < ω, (Iio o),
+    (∀ i : Ordinal.{u}, (hi : i < ω) → P (f i hi) (f (i + 1) (omega_isLimit.2 i hi)))
+    ∧ (∀ i j, (hi : i < ω) → (hj : j < ω) → (i < j) → f i hi < f j hj)
+    ∧ r < f 0 omega_pos := by
+  have oLim : o.IsLimit := ⟨(zero_lt_one.trans onelto).ne.symm, fun a alto ↦ by
+    by_cases h : a = 0
+    · exact h ▸ succ_zero ▸ onelto
+    rcases hP a alto (Ordinal.pos_iff_ne_zero.mpr h) with ⟨w, hw⟩
+    exact lt_of_le_of_lt (succ_le_iff.mpr hw.2.1) hw.1⟩
+  let P' : Ordinal → Ordinal → Prop := fun p q ↦ p = 0 ∨ P p q
+  have hP' : ∀ p < o, ∃ q < o, (p < q ∧ P' p q) := fun p plto ↦ by
+    by_cases h : p = 0
+    · use 1; use succ_zero ▸ (oLim.succ_lt oLim.pos); use h ▸ zero_lt_one; exact Or.inl h
+    convert hP p plto (Ordinal.pos_iff_ne_zero.mpr h) using 1
+    simp_all only [false_or, P']
+  rcases exists_omega_seq_succ_prop (zero_lt_one.trans onelto) hP' rlto with ⟨f, hf⟩
+  use f
+  refine' ⟨fun i hi ↦ _, hf.2⟩
+  have := hf.1 i hi
+  have rltf0 := hf.2.2
+  by_cases hi' : i = 0
+  · subst hi'
+    exact this.resolve_left (pos_of_gt rltf0).ne.symm
+  · have rltfi := rltf0.trans <| hf.2.1 0 i omega_pos hi (Ordinal.pos_iff_ne_zero.mpr hi')
+    exact this.resolve_left (pos_of_gt rltfi).ne.symm
+
+/-- If between every 2 consecutive elements of an increasing `δ`-sequence
+  there is an element of `C`, and `δ` is a limit ordinal,
+  then the supremum of the sequence is an accumulation point of `C`. -/
+theorem isAcc_bsup_of_between {δ : Ordinal} (C : Set Ordinal) (δLim : δ.IsLimit)
+    (s : Π o < δ, Ordinal) (sInc : ∀ o, (h : o < δ) → s o h < s (o + 1) (δLim.succ_lt h))
```

Comment:
I kind of defaulted to using `Π o < δ, Ordinal` because that's how fundamental sequences are defined, but `Iio δ` is probably better in many places (maybe including fundamental sequences, but that ship has sailed). I'll change it and work with `sup` instead of `bsup`.